numpy/__init__.pyi

```diff
@@ -938,9 +938,9 @@ class flatiter(Generic[_NdArraySubClass]):
         value: Any,
     ) -> None: ...
     @overload
-    def __array__(self: flatiter[ndarray[Any, _DType]], dtype: None = ..., /) -> ndarray[Any, _DType]: ...
+    def __array__(self: flatiter[ndarray[_ShapeType, _DType]], dtype: None = ..., /) -> ndarray[_ShapeType, _DType]: ...
     @overload
-    def __array__(self, dtype: _DType, /) -> ndarray[Any, _DType]: ...
+    def __array__(self: flatiter[ndarray[_ShapeType, Any]], dtype: _DType, /) -> ndarray[_ShapeType, _DType]: ...
```

Comment:
This doesn't seem right.  `flatiter` returns a 1-D object, while `self` can have any number of dimensions?

Reply:
How about using a `_Shape1DType = TypeVar(..., bound=tuple[int])` then?
This way, the shape could still be fixed like `tuple[Literal[42]]` or unspecified (yet 1d) as `tuple[int]`.

Reply:
Actually now that I think about it, that might not be a good idea: 
Because at the moment, the `ndarray` shape type param has no upper bound, so that wouldn't be backwards compatible

Reply:
I honestly don't know and will trust you if you find a solution.  It just seemed wrong to pass on the shape from the array.
If you can encode 1-D in a compatible way, that is good of course.

Reply:
Of nevermind, your point just landed:

```pycon
>>> import numpy as np
>>> a = np.arange(8).reshape(4, -1)
>>> np.array(a.flat).shape
(8,)
```

It shouldn't be too difficult to deal with this, by adding a specific overload for `ndarray[_FlatShapeType, _DType]` with `_FlatShape: tuple[int]`, and otherwise returning a shape type of `tuple[int]`

Reply:
I opted to leave the shape `Any` in case of unknown shape of the underlying array, since shape types aren't officially required to be integer tuples at the moment.

But this now correctly passes the shape type when the underlying array has a shape type like `tuple[int]`, and added some tests for it.